### DISCUSSION TITLE
std::flat_map allocator-extended copy/move constructor

### DISCUSSION THREAD

**Discussion by duanqn** [microsoft/STL#4779] [Q&A] [answered]
Hi,
I am trying to compile the following code and it fails

```
{
    using flat_map_t = std::flat_map<int, int, std::less<void>, std::vector<int, std::allocator<int>>, std::vector<int, std::allocator<int>>>;
    flat_map_t map0;
    flat_map_t map1(map0);
    auto allocator = std::allocator<int>();
    flat_map_t map2(map0, allocator);
    flat_map_t map3(std::move(map0));
    flat_map_t map4(std::move(map1), allocator);
}
```

msvc compiler version: 19.41.33923
STL: 73f969466bdcd4ce5fae3365ce759b29adf45744

Error message: C2448 'std::flat_map<int,int,std::less<void>,std::vector<int,std::allocator<int>>,std::vector<int,std::allocator<int>>>::_Key_compare': cannot access private member declared in class 'std::flat_map<int,int,std::less<void>,std::vector<int,std::allocator<int>>,std::vector<int,std::allocator<int>>>'

Apparently, the compiler is complaining about this part

```
    template <_Valid_allocator_for_flat_map<key_container_type, mapped_container_type> _Allocator>
    _Flat_map_base(const _Derived& _Other, const _Allocator& _Alloc)
        : _Key_compare(_Other._Key_compare),
          _Data{.keys = _STD make_obj_using_allocator<key_container_type>(_Alloc, _Other._Data.keys),
              .values = _STD make_obj_using_allocator<mapped_container_type>(_Alloc, _Other._Data.values)} {}
```

used here

`flat_map_t map2(map0, allocator);`

According to the compiler, we are not allowed to access `_Derived::Key_compare` and `_Derived::Data` from `_Flat_map_base` because they are declared with `using _MyBase::Data` under the `private` section in `flat_map`.
However, there is no compilation error for the regular copy constructor:

```
_Flat_map_base(const _Derived& _Other) : _Key_compare(_Other._Key_compare), _Data(_Other._Data) {}
```

Is this a compiler bug, or an implementation bug of <flat_map>? How should we solve it?

**Comment by duanqn** [comment]
@frederick-vs-ja  @AlexGuteniev   FYI

**Comment by frederick-vs-ja** [comment, marked as answer]
I think this is a bug. `_Flat_map_base` doesn't have access to the members in derived classes (where the access control is changed to private).

I'll fix this by `static_cast`-ing the parameters to (`const`) `_Flat_map_base&` first.
